File: 24/24.py
import sys
import random


def reduce(numbers, exps):

	firstNumber = []
	firstNumberExp = []

	# if numbers[0] > numbers[1]:
	# 	# exchange the order of the first two
	# 	num = numbers[0]
	# 	numbers[0] = numbers[1]
	# 	numbers[1] = num
	# 	exp = exps[0]
	# 	exps[0] = exps[1]
	# 	exps[1] = exp	

	# + 
	firstNumber.append(numbers[0]+numbers[1])
	firstNumberExp.append("("+exps[0]+"+"+exps[1]+")")
	
	# -
	if numbers[0]-numbers[1] >= 0 :
		firstNumber.append(numbers[0]-numbers[1])
		firstNumberExp.append("("+exps[0]+"-"+exps[1]+")")
	else :	
		firstNumber.append(numbers[1]-numbers[0])
		firstNumberExp.append("("+exps[1]+"-"+exps[0]+")")
	
	# * 
	firstNumber.append(numbers[0]*numbers[1])
	firstNumberExp.append("("+exps[0]+"*"+exps[1]+")")
	
	# /
	if numbers[0] != 0:
		firstNumber.append(numbers[1]/numbers[0])
		firstNumberExp.append("("+exps[1]+"/"+exps[0]+")")
		
	if numbers[1] != 0:
		firstNumber.append(numbers[0]/numbers[1])
		firstNumberExp.append("("+exps[0]+"/"+exps[1]+")")
	

	newList = []
	for i in range(len(firstNumber)): 
		list = []
		list.append(firstNumber[i])
		list.extend(numbers[2:])
		exp = []
		exp.append(firstNumberExp[i])
		exp.extend(exps[2:])
		newList.append({"l":list, "e":exp})


	return newList


def reduceAll(list):

	newList = []
	for element in list: 
		numbers = element["l"]
		exps = element["e"]
		s = len(numbers)
		for i in range(s):
			for j in range(s):
				if i != j:
					list4=[]
					list4Exp = []
					list4.append(numbers[i])
					list4Exp.append(exps[i])
					list4.append(numbers[j])
					list4Exp.append(exps[j])
					for k in range(s):
						if k != i and k != j:
							list4.append(numbers[k])
							list4Exp.append(exps[k])
					obj = reduce(list4,list4Exp)
					# print(obj)
					newList.extend(obj)
					

	# remove redundency
	hash = set()
	ret = []
	for obj in newList:
		exp = obj["e"]
		key = str(sorted(exp))
		#print("key:"+key)
		if  key not in hash:
			ret.append(obj)
			hash.add(key)

	return ret

def solve(nums):
	print("Solving:"+str(nums))
	size = len(nums)
	exps = []
	for i in range(size):
		exps.append(str(nums[i]))
	# exps = [str(nums[0]), str(nums[1]), str(nums[2]), str(nums[3])]
	list = [{"l":nums, "e":exps}]
	for i in range(size-1):
		list = reduceAll(list)
		# print(list)
	count = 0
	for obj in list: 
		if obj["l"][0] == 24: 
			print(obj["e"][0])
			count = count + 1	
	print("# of solutions: "+str(count))
	return count > 0



def solveAll():
	solvable = 0
	unsolvable = 0
	for i in range(1,11):
		for j in range(i,11):
			for k in range(j,11):
				for l in range(k,11):
					numbers = [i,j,k,l]
					#exp = [str(i), str(j), str(k), str(l)]
					if solve(numbers):
						solvable += 1
					else: 
						unsolvable +=1

	print("Total solvable: "+str(solvable))
	print("Total unsolvable: "+str(unsolvable))
	print("Solvable rate: "+ str(solvable/(solvable+unsolvable)))



size = len(sys.argv) - 1

if size > 1 : 
# input numbers
	numbers = []
	for i in range(1,size+1):
		 numbers.append(int(sys.argv[i]))
	solve(numbers)	
elif size == 1:
	choices = [1,2,3,4,5,6,7,8,9,10]
	numbers = [random.choice(choices),random.choice(choices),random.choice(choices),random.choice(choices)]
	print(numbers)
	input("Enter to see the solutions:")
	solve(numbers)

else:

	solveAll()

File: 24/24.js
var message ="";		
var reduce = function(numbers, exps) {

	//message += "<br/><br/>reduce: {["+numbers.join(",")+"] ["+exps.join(",")+"]}";
	var firstNumber = [];
	var firstNumberExp = [];

	// + 
	firstNumber.push(numbers[0]+numbers[1]);
	firstNumberExp.push("("+exps[0]+"+"+exps[1]+")");
	
    // -
	if (numbers[0]-numbers[1] >= 0) {
		firstNumber.push(numbers[0]-numbers[1]);
		firstNumberExp.push("("+exps[0]+"-"+exps[1]+")");
	}
	else if (numbers[0] !== numbers[1]) {	
		firstNumber.push(numbers[1]-numbers[0]);
		firstNumberExp.push("("+exps[1]+"-"+exps[0]+")");
	}
	// * 
	firstNumber.push(numbers[0]*numbers[1]);
	firstNumberExp.push("("+exps[0]+"*"+exps[1]+")");
	
	// /
	if (numbers[0] !== 0) {
		firstNumber.push(numbers[1]/numbers[0]);
		firstNumberExp.push("("+exps[1]+"/"+exps[0]+")");
	}
	if (numbers[1] !== 0 && numbers[1] !== numbers[0]) {
		firstNumber.push(numbers[0]/numbers[1]);
		firstNumberExp.push("("+exps[0]+"/"+exps[1]+")");
	}

	var newList = [];
	for (var i=0;i<firstNumber.length;i++) { 
		var list = [];
		list.push(firstNumber[i]);
		var exp = [];
		exp.push(firstNumberExp[i]);
		for (var j = 2; j< numbers.length; j++) {
		    list.push(numbers[j]);
		    exp.push(exps[j]);
		}
		//list.push(numbers[3]);
		//list.concat(numbers.slice(2,numbers.length-2));
        //exp.concat(exps.slice(2));
		var obj ={};
		obj.l = list;
		obj.e = exp;
		// message+= list.join(",");
		// message+=" ";
		// message+= exp.join(",");
		// message+=" ";
		newList.push(obj);
		
    }
    //message += "List size:"+newList.length;
	return newList;   
};

var reduceAll = function(list) {

	//message += "<br/><br/>reduceAll"
	var newList = [];
	for (var index = 0;index < list.length; index++) {
	    var element = list[index];

		var numbers = element.l;
		//message = numbers.toString();
		var exps = element.e;
        //message += numbers.join(",");
        //message += "<br/><br/>reducing "+numbers.join(",");
		var s = numbers.length; 
		var hash = new Set();

		for (var i =0;i < s; i++) {
			for (var j = 0; j <s; j++) {
				if (i !== j) {
					var list4=[];
					var list4Exp = [];
					list4.push(numbers[i]);
					list4Exp.push(exps[i]);
					list4.push(numbers[j]);
					list4Exp.push(exps[j]);
					if (s > 2) {
						for (var k = 0; k < s; k++) {
							if (k !== i && k !== j) {
								list4.push(numbers[k]);
								list4Exp.push(exps[k]);
							}
						}
					}
					//message += list4Exp.join(",");
					var key = list4.join(",");
					if (!hash.has(key)) {
						var obj = reduce(list4,list4Exp);
						//message+=obj.length;
						//newList.concat(obj);
						for ( var l = 0;l<obj.length;l++) {
							newList.push(obj[l]);
						}
						hash.add(key);
					}
				}
			}
		}
					
	}
	//message +="<br/><br/>reduceAll return size: "+newList.length;
	// return newList;
	
	// remove redundency
	var hash = new Set();
	var ret = [];
	for (var i = 0;i<newList.length;i++) {
		var obj = newList[i];
		var numbers = obj.e;
		var key = numbers.slice().sort().toString();
		//print("key:"+key)
		if  (!hash.has(key)) {
			ret.push(obj);
			hash.add(key);
		}
	}
	return ret;
};

var solve = function(nums) {
    //message += "<br/> Solving"+nums.join(",");
	var size = nums.length;
	var exps = [];
	for (var i = 0; i < size ; i++) {
		exps.push(nums[i].toString());
	}
	var obj = {};
	obj.l = nums;
	obj.e = exps;
	//message = obj.e.toString();
	var list = [obj];
	for (var i = 0; i< size-1; i++) {
		list = reduceAll(list);
	}
	
    // message = list.toString();
	
	var count = 0;
	for (var i = 0; i<list.length;i++) { 
		var obj = list[i];
		if (obj.l[0] === 24) { 
			message += obj.e[0];
			message += ", "
			count = count + 1;
		}
	}
	message += "<br/><br/># of Solutions: "+count.toString();

};

var solveIt = function() {
    //reduceAll({"l":[1,2,3,4],"e":["1","2","3","4"]});
    var n1 = parseInt(document.getElementById("n1").value);
    var n2 = parseInt(document.getElementById("n2").value);
    var n3 = parseInt(document.getElementById("n3").value);
    var n4 = parseInt(document.getElementById("n4").value);
    message = "";
    solve([n1,n2,n3,n4]);

    document.getElementById("message").innerHTML = message;
};
